Fix LastWriteTime and LastAccessTime of a symlink on Windows and Unix

hamarb123 opened the pull request:
Fixes #38824.
Fixes setting the LastWriteTime and the LastAccessTime on a symlink on Windows and Unix.
Somewhat reliant on https://github.com/dotnet/runtime/pull/49555.
Reverts the Windows changes of the commit https://github.com/dotnet/runtime/commit/a617a01195b4dad3cb5f300962ad843b46d4f175 (on the https://github.com/dotnet/runtime/pull/49555 PR) to fix #38824 and for Unix: https://github.com/hamarb123/runtime/commit/f84293583bb0c8e1060bdacfb80158a6c64c5fbd.

hamarb123 committed:
Implement most of the fix for #38824

Reverts the changes in the seperate PR https://github.com/dotnet/runtime/commit/a617a01195b4dad3cb5f300962ad843b46d4f175 to fix #38824.
Does not re-enable the test because that relies on #49555, will add a seperate commit to whichever is merged last to enable the SettingUpdatesPropertiesOnSymlink test.

carlossanlop reviewed (commented):
The change looks good so far, @hamarb123, thanks for the fix.

As mentioned in your description, unit tests should be included, so please ping me when you add them.

jeffhandley commented:
Hi @hamarb123 -- it looks like the we were waiting for some unit tests to be added. Let us know if you need any help or guidance on that. Thanks!

hamarb123 commented:
@jeffhandley, unit tests are in #49555, they just need to be enabled on Windows when #49555 is merged.

jeffhandley commented:
Ah; I'm sorry I overlooked that @hamarb123!

jeffhandley commented:
@Jozkee This PR is assigned to you for follow-up/decision before the RC1 snap.

hamarb123 commented:
I'm separating out the symlink code into this PR for Unix from #49555 as per https://github.com/dotnet/runtime/pull/49555#issuecomment-904413721.

hamarb123 commented:
I'm not sure how to make this PR include the changes of the other PR without having it as seperate commits that need to also be merged, so for now I'll just remove the relevant code from #49555 and add it here once it's merged. If it is possible, please let me know how I can do it because this PR needs to change files added in #49555.

danmoseley commented:
I see this is blocked on https://github.com/dotnet/runtime/pull/49555. Perhaps we/you can reopen this PR when this is unblocked.

hamarb123 commented:
Hi, please reopen this PR. Thanks!

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into fixfor38824

hamarb123 committed:
Most of the code for PR #52639 to fix #38824

• Deal with merge conflicts
• Add FSOPT_NOFOLLOW for macOS and use it in setattrlist
• Use AT_SYMLINK_NOFOLLOW with utimensat (note: there doesn't seem to be an equivalent for utimes)
• Add SettingUpdatesPropertiesOnSymlink test to test if it actually sets it on the symlink itself (to the best that we can anyway ie. write + read the times)
• Specify FILE_FLAG_OPEN_REPARSE_POINT for CreateFile in FileSystem.Windows.cs (is there any other CreateFile calls that need this?)

hamarb123 committed:
Remove additional FILE_FLAG_OPEN_REPARSE_POINT

I added FILE_FLAG_OPEN_REPARSE_POINT before and it was then added in another PR, this removes the duplicate entry.

hamarb123 committed:
Add missing override keywords

Add missing override keywords for abstract members in the tests.

hamarb123 committed:
Fix access modifiers

Fix access modifiers for tests - were meant to be protected rather than public

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into fixfor38824

jeffhandley commented:
We're inconsistent with how we use the assignees field and I should have commented what my intent was. My intent is for you (as the author) to drive this PR through to completion but to have @Jozkee assigned as the team member accountable for ensuring we're supporting you as needed. You're still assigned the issue though. Is that OK?

(We've been discussing some better practices for ensuring we're not leaving contributors hanging for long periods of time).

hamarb123 committed:
Add more symlink tests, rearrange files

• Move symlink creation api to better spot in non-base files
• Add IsDirectory property for some of the new tests
• Change abstract symlink api to CreateSymlink that accepts path and target
• Create a CreateSymlinkToItem method that creates a symlink to an item that may be relative that uses the new/modified abstract CreateSymlink method
• Add SettingUpdatesPropertiesCore to avoid code duplication
• Add tests for the following variants: normal/symlink, target exists/doesn't exist, absolute/relative target
• Exclude nonexistent symlink target tests on Unix for Directories since they are counted as files

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into fixfor38824

hamarb123 committed:
Fix return type of CreateSymlink in File/GetSetTimes.cs

hamarb123 committed:
Remove browser from new symlink tests as it doesn't support creation of symlinks

Remove browser from new symlink tests as it doesn't support creation of symlinks

jozkee reviewed (commented):
@hamarb123 can you please address using `lutimes` instead of `utimes` as pointed out here https://github.com/dotnet/runtime/pull/52639/files#r737921487?

hamarb123 committed:
Use lutimes, improve code readability, simplify tests

• Use lutimes when it's available
• Extract dwFlagsAndAttributes to a variable
• Use same year for all tests
• Checking to delete old symlink is unnecessary, so don't
• Replace var with explicit type

hamarb123 committed:
Change year in test to 2014 to reduce diff

jozkee reviewed (approved):
Otherwise, LGTM. Thanks!

hamarb123 committed:
Rename symlink tests, use 1 core symlink times function, and check that target times don't change

Rename symlink tests, use 1 core symlink times function, and check that target times don't change

hamarb123 committed:
Inline RunSymlinkTestPart 'function'

Inline RunSymlinkTestPart 'function' so that the code can be reordered so the access time test can be valid.

hamarb123 committed:
Share CreateSymlinkToItem call in tests and update comment for clarity

hamarb123 committed:
Update symlink time tests

• Make SettingUpdatesPropertiesOnSymlink a theory
• Remove special case for Unix due to https://github.com/dotnet/runtime/pull/52639#discussion_r739009259 (will revert if fails)
• Rename isRelative to targetIsRelative for clarity

hamarb123 committed:
Remove unnecessary Assert.All

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,95 @@ public void SettingUpdatesProperties()
+            // The symlink test is [0] of the Assert.All, and the target check is [1].
+
+            T item = CreateSymlinkToItem(target, isRelative);
+            if (!targetExists) SettingUpdatesPropertiesCore(item);
```
> It would be good to add some randomization here. One setter could cause subsequent getters to succeed even if their setter is broken.

Are you referring to randomize the date that is set in `SettingUpdatesPropertiesCore`, or randomize the dates of the target, or both?

Jozkee replied:
```suggestion
            if (!targetExists) 
            {
                SettingUpdatesPropertiesCore(item);
            }
```

hamarb123 replied:
This is in the next commit.

hamarb123 replied:
> Are you referring to randomize the date that is set in `SettingUpdatesPropertiesCore`, or randomize the dates of the target, or both?

I think they're referring to randomising both, but it's unnecessary since you said path names are randomised.

tmds replied:
> Are you referring to randomize the date that is set in SettingUpdatesPropertiesCore, or randomize the dates of the target, or both?

`SettingUpdatesPropertiesCore`. The getters/setters run on the same path, so the path being unique doesn't detect issues where one setter affects multiple getters.

hamarb123 replied:
> `SettingUpdatesPropertiesCore`. The getters/setters run on the same path, so the path being unique doesn't detect issues where one setter affects multiple getters.

There's a different test for that (for normal files). It's assumed that it will also work on symlinks if times are set correctly on symlinks rather than the targets (which this test checks), since there is no code that currently, and shouldn't ever probably be any time setting/getting code that treats symlinks differently to normal files (it would likely fail this test also if it did anyway). Note the condition which is most likely for one setter affecting multiple getters is "when the modification time is less than the creation time" (if code was done from scratch for macOS without reading some of the comments would be when this is an issue), so simply testing if particular setters change more than 1 getter isn't good enough.
I do think it's probably unnecessary, but if @Jozkee wants, I could also implement the `SettingUpdatesPropertiesAfterAnother` test for symlinks, but again, I think it's probably overkill since it's being checked for normal files.

tmds replied:
My comment is not specific to this PR. It applies also to the `SettingUpdatesProperties` test as it was.

hamarb123 replied:
See the `SettingUpdatesPropertiesAfterAnother` test.

tmds replied:
Ah, yes, `SettingUpdatesPropertiesAfterAnother` will catch it.

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
+            else
+            {
+                // Ensure that we have the latest times (access time could be changed by creating the symlink)
+                FileSystemInfo fsi = target as FileSystemInfo;
```
This will be `null` when `T` is `string`. To avoid that you can add an abstract method `GetFileSystemInfo(T item)` and define how to get a FileSystemInfo for each derived class.

hamarb123 replied:
Hi, the point of this code is to call `FileSystemInfo.Refresh` to invalidate the times cache before getting the times. If it uses `File` or `Directory`, it will not be cached (right?). Hence if creating a link can change these times, the cache needs to be refreshed. Also, it definitely needs to be refreshed after running the test on the link since that's the point of the extra check.

hamarb123 replied:
@tmds says it will work without the first one, so I'm only keeping the 2nd one (which definitely is necessary, else it'll return the cached times). https://github.com/dotnet/runtime/pull/52639#discussion_r739761174

hamarb123 commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
+            {
+                // Ensure that we have the latest times (access time could be changed by creating the symlink)
+                FileSystemInfo fsi = target as FileSystemInfo;
+                fsi?.Refresh();
```
I'm not familiar with when `atime` updates. When a symlink is created on Unix, it checks if the target exists and then if it's a file or a directory. To me this sounds like accessing, please correct me if there's no chance of `atime` updating. @Jozkee, do you know if `atime` might update?

tmds replied:
`atime` changes when a file is opened for reading. That does not happen here.

hamarb123 replied:
👍 ok, I'll remove the initial refresh. If the assert fails (which it shouldn't), I'll put it back.

hamarb123 commented on src/libraries/System.IO.FileSystem/tests/Directory/GetSetTimes.cs:
```diff
@@ -9,6 +9,10 @@ public class Directory_GetSetTimes : StaticGetSetTimes
 
+        protected override string CreateSymlink(string path, string pathToTarget) => Directory.CreateSymbolicLink(path, pathToTarget).FullName;
+
+        protected override bool IsDirectory => true;
```
In the next commit.

hamarb123 commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
+                fsi?.Refresh();
+
+                // We only use UTC times since checking the others is unnecessary.
+                IEnumerable<TimeFunction> timeFunctionsUtc = TimeFunctions(requiresRoundtripping: true).Where((f) => f.Kind == DateTimeKind.Utc);
```
In the next commit.

hamarb123 commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
+
+            // When the target exists, we want to verify that its times don't change.
+
+            T item = CreateSymlinkToItem(target, targetIsRelative);
```
In the next commit.

hamarb123 commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +80,60 @@ public void SettingUpdatesProperties()
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(true, true)]
+        public void SettingUpdatesPropertiesOnSymlink(bool targetIsRelative, bool targetExists)
```
I thought it might be a good test case and @Jozkee agreed, it is probably unnecessary though. https://github.com/dotnet/runtime/pull/52639#discussion_r736013152

tmds replied:
@Jozkee the additional argument checks whether both relative and full paths work as the symlink target path. It's not related to the functionality that gets tested here.

hamarb123 replied:
The reason that we included it here is we're checking if it changes the date on the target file as well, so all variants of the 'target' should be included IMO (which is what it currently is: relative/absolute, exists/doesn't, file/folder, info/static).

tmds replied:
> all variants of the 'target' should be included IMO

Another variant would be using multiple links to get to the target.

Because the target gets resolved by the OS, I don't see the need use different target paths.

I'm ok if you want to keep it.

hamarb123 replied:
> Another variant would be using multiple links to get to the target.

That's true. I don't see it as necessary personally.

> Because the target gets resolved by the OS, I don't see the need use different target paths.

Probably fair. I'd rather it be there, if @Jozkee wants I can remove it.

> I'm ok if you want to keep it.

👍 thanks

Jozkee replied:
What @tmds says makes sense. You can remove `targetIsRelative` if you want.

hamarb123 committed:
Changes to SettingUpdatesPropertiesOnSymlink test

• Rename item field to link field
• Don't use if one-liner
• Use all time functions since only using UTC isn't necessary
• Remove the now-defunct IsDirectory property since we aren't checking it anymore

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into fixfor38824

hamarb123 committed:
Remove unnecessary fsi.Refresh()

• Remove unnecessary fsi.Refresh() since atime is only updated when reading a file

tmds reviewed (approved):
lgtm, thanks!

Jozkee commented on src/libraries/Native/Unix/System.Native/pal_time.c:
```diff
@@ -33,7 +33,15 @@ int32_t SystemNative_UTimensat(const char* path, TimeSpec* times)
-    while (CheckInterrupted(result = utimensat(AT_FDCWD, path, updatedTimes, 0)));
+    while (CheckInterrupted(result = utimensat(AT_FDCWD, path, updatedTimes, AT_SYMLINK_NOFOLLOW)));
+#elif HAVE_LUTIMES
+    struct timeval updatedTimes[2];
+    updatedTimes[0].tv_sec = (long)times[0].tv_sec;
+    updatedTimes[0].tv_usec = (int)times[0].tv_nsec / 1000;
+
+    updatedTimes[1].tv_sec = (long)times[1].tv_sec;
+    updatedTimes[1].tv_usec = (int)times[1].tv_nsec / 1000;
+    while (CheckInterrupted(result = lutimes(path, updatedTimes)));
```
You can re-use this code for lutimes and utimes i.e: the `#elif` can be shrinked to lutimes vs utimes.

hamarb123 replied:
I've combined. Let me know if you want it combined differently (ie. #if ing a whole line rather than a function name).

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:
```diff
@@ -71,6 +79,56 @@ public void SettingUpdatesProperties()
+                SettingUpdatesPropertiesCore(link);
+
+                // Ensure that we have the latest times.
+                if (target is FileSystemInfo fsi) fsi.Refresh();
```
```suggestion
                if (target is FileSystemInfo fsi)
                { 
                    fsi.Refresh();
                }
```

hamarb123 replied:
Done

hamarb123 committed:
Updates to test and pal_time.c

• Remove targetIsRelative cases
• Multi-line if statement
• Combine HAVE_LUTIMES and #else conditions to allow more code charing

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into fixfor38824

hamarb123 committed:
Remove trailing space

Jozkee commented:
Breaking change doc issue: https://github.com/dotnet/docs/issues/31483
